node/subsystem/src/util.rs

```diff
@@ -0,0 +1,563 @@
+/// Other messages are passed along to and from the job via the overseer to other
+/// subsystems.
+pub trait JobTrait {
+	/// Message type to the job. Typically a subset of AllMessages.
+	type ToJob: 'static + ToJobTrait + Send;
+	/// Message type from the job. Typically a subset of AllMessages.
+	type FromJob: 'static + Into<AllMessages> + Send;
+	/// Job runtime error.
+	type Error: std::fmt::Debug;
+	/// Extra arguments this job needs to run properly.
+	///
+	/// If no extra information is needed, it is perfectly acceptable to set it to `()`.
+	type RunArgs: 'static + Send;
+
+	/// Name of the job, i.e. `CandidateBackingJob`
+	const NAME: &'static str;
+
+	/// Run a job for the parent block indicated
+	fn run(
+		parent: Hash,
+		run_args: Self::RunArgs,
+		rx_to: mpsc::Receiver<Self::ToJob>,
+		tx_from: mpsc::Sender<Self::FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>>;
+
+	/// Handle a message which can't be dispatched to a particular job
+	///
+	/// By default, this is implemented with a NOP function. However, if
+	/// ToJob occasionally has messages which do not correspond to a particular
+	/// parent relay hash, then this function will be spawned as a one-off
+	/// task to handle those messages.
+	// TODO: the API here is likely not precisely what we want; figure it out more
+	// once we're implementing a subsystem which actually needs this feature.
+	// In particular, we're quite likely to want this to return a future instead of
+	// interrupting the active thread for the duration of the handler.
+	fn handle_unhashed_msg(_msg: Self::ToJob) -> Result<(), Self::Error> {
+		Ok(())
+	}
+}
+
+pub struct Jobs<Spawner, Job: JobTrait> {
+	spawner: Spawner,
+	running: HashMap<Hash, JobHandle<Job::ToJob>>,
+	outgoing_msgs: StreamUnordered<mpsc::Receiver<Job::FromJob>>,
+	job: std::marker::PhantomData<Job>,
+}
+
+impl<Spawner: Spawn, Job: JobTrait> Jobs<Spawner, Job> {
+	/// Create a new Jobs manager which handles spawning appropriate jobs.
+	pub fn new(spawner: Spawner) -> Self {
+		Self {
+			spawner,
+			running: HashMap::new(),
+			outgoing_msgs: StreamUnordered::new(),
+			job: std::marker::PhantomData,
+		}
+	}
+
+	/// Spawn a new job for this `parent_hash`, with whatever args are appropriate.
+	fn spawn_job(&mut self, parent_hash: Hash, run_args: Job::RunArgs) -> Result<(), Error> {
+		let (to_job_tx, to_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (from_job_tx, from_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (finished_tx, finished) = oneshot::channel();
+
+		let (future, abort_handle) = future::abortable(async move {
+			if let Err(e) = Job::run(parent_hash, run_args, to_job_rx, from_job_tx).await {
+				log::error!(
+					"{}({}) finished with an error {:?}",
+					Job::NAME,
+					parent_hash,
+					e,
+				);
+			}
+		});
+
+		// discard output
+		let future = async move {
+			let _ = future.await;
+			let _ = finished_tx.send(());
+		};
+		self.spawner.spawn(future)?;
```

Comment:
this returns a future, which some times might come in handy when collecting jobs, could we return the future?